src/coord/src/coord/prometheus.rs

```diff
@@ -0,0 +1,151 @@
+    registry: &'a Registry,
+    command_rx: std::sync::mpsc::Receiver<ScraperMessage>,
+    internal_tx: UnboundedSender<super::Message>,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum ScraperMessage {
+    Shutdown,
+}
+
+fn convert_metrics_to_rows<'a, M: IntoIterator<Item = &'a prometheus::proto::Metric>>(
+    kind: MetricType,
+    metrics: M,
+) -> Vec<MetricReading> {
+    metrics
+        .into_iter()
+        .flat_map(|m| {
+            use MetricType::*;
+            let labels = m
+                .get_label()
+                .into_iter()
+                .map(|pair| (pair.get_name().to_owned(), pair.get_value().to_owned()))
+                .collect::<Vec<(String, String)>>();
+            match kind {
+                COUNTER => Some(MetricReading::new(
+                    labels,
+                    MetricValue::Value(m.get_counter().get_value()),
+                )),
+                GAUGE => Some(MetricReading::new(
+                    labels,
+                    MetricValue::Value(m.get_gauge().get_value()),
+                )),
+                HISTOGRAM => {
+                    let histo = m.get_histogram();
+                    Some(MetricReading::new(
+                        labels,
+                        MetricValue::Histogram {
+                            sum: histo.get_sample_sum(),
+                            total_count: histo.get_sample_count() as i64,
+                            counts: histo
+                                .get_bucket()
+                                .iter()
+                                .map(|b| b.get_cumulative_count() as i64)
+                                .collect(),
+                            bounds: histo
+                                .get_bucket()
+                                .iter()
+                                .map(|b| b.get_upper_bound())
+                                .collect(),
+                        },
+                    ))
+                }
+                // TODO: destructure summaries once we have any.
+                _ => None,
+            }
+        })
+        .collect()
+}
+
+impl<'a> Scraper<'a> {
+    pub fn new(
+        interval: Duration,
+        retain_for: Duration,
+        registry: &'a Registry,
+        command_rx: std::sync::mpsc::Receiver<ScraperMessage>,
+        internal_tx: UnboundedSender<super::Message>,
+    ) -> Self {
+        Scraper {
+            interval,
+            retain_for,
+            registry,
+            command_rx,
+            internal_tx,
+        }
+    }
+
+    /// Run forever: Scrape the metrics registry once per interval, telling the coordinator to
+    /// insert the values and meta-info in internal tables.
+    pub fn run(&mut self) {
+        let retain_for = self.retain_for.as_millis() as u64;
+        loop {
+            thread::sleep(self.interval);
+            let timestamp: Timestamp = UNIX_EPOCH
+                .elapsed()
+                .expect("system clock before 1970")
```

Comment:
nit: normally we write what we do expect, not what caused the expectation to fail: `.expect("system clock after 1970")`, and the same in the expect below.

Reply:
Haha, I think I cargo-culted that from elsewhere, but happy to fix here (: